File: frontend/src/features/tasks/task.api.ts
import { api, query } from '../../lib/api/client'
import type { Page, Project, Subtask, Tag, Task, TaskInput } from './task.types'
export const listTasks = (params: Record<string, string | number | undefined> = {}) =>
  api<Page<Task>>(`/v1/tasks${query(params)}`)
export const getTask = (id: string) => api<Task>(`/v1/tasks/${id}`)
export const createTask = (input: TaskInput) =>
  api<Task>('/v1/tasks', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
export const updateTask = (id: string, input: Partial<TaskInput>) =>
  api<Task>(`/v1/tasks/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
export const setCompletion = (id: string, completed: boolean) =>
  api<Task>(`/v1/tasks/${id}/completion`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ completed }),
  })
export const deleteTask = (id: string) => api<void>(`/v1/tasks/${id}`, { method: 'DELETE' })
export const createSubtask = (
  taskId: string,
  title: string,
  completed = false,
  position = 0,
) =>
  api<Subtask>(`/v1/tasks/${taskId}/subtasks`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ title, completed, position }),
  })
export const updateSubtask = (
  taskId: string,
  subtaskId: string,
  input: { title?: string; completed?: boolean; position?: number },
) =>
  api<Subtask>(`/v1/tasks/${taskId}/subtasks/${subtaskId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
export const deleteSubtask = (taskId: string, subtaskId: string) =>
  api<void>(`/v1/tasks/${taskId}/subtasks/${subtaskId}`, { method: 'DELETE' })
export const listProjects = () => api<Page<Project>>('/v1/projects')
export const createProject = (name: string, color?: string) =>
  api<Project>('/v1/projects', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name, color: color || null }),
  })
export const listTags = () => api<Page<Tag>>('/v1/tags')
export const createTag = (name: string) =>
  api<Tag>('/v1/tags', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })

File: android/app/src/main/java/com/tasker/android/remote/api/TaskApi.kt
package com.tasker.android.remote.api

import com.tasker.android.remote.dto.CompletionRequest
import com.tasker.android.remote.dto.ItemsResponse
import com.tasker.android.remote.dto.ProjectCreateRequest
import com.tasker.android.remote.dto.ProjectDto
import com.tasker.android.remote.dto.TagCreateRequest
import com.tasker.android.remote.dto.TagDto
import com.tasker.android.remote.dto.TaskCreateRequest
import com.tasker.android.remote.dto.TaskDto
import com.tasker.android.remote.dto.TaskUpdateRequest
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

interface TaskApi {

    // ── Tasks ──────────────────────────────────────────────────────

    @GET("v1/tasks")
    suspend fun listTasks(
        @Query("status") status: String = "all",
        @Query("limit") limit: Int = 1000,
        @Query("project_id") projectId: String? = null,
        @Query("q") query: String? = null,
    ): ItemsResponse<TaskDto>

    @POST("v1/tasks")
    suspend fun createTask(@Body body: TaskCreateRequest): TaskDto

    @GET("v1/tasks/{id}")
    suspend fun getTask(@Path("id") id: String): TaskDto

    @PATCH("v1/tasks/{id}")
    suspend fun updateTask(@Path("id") id: String, @Body body: TaskUpdateRequest): TaskDto

    @DELETE("v1/tasks/{id}")
    suspend fun deleteTask(@Path("id") id: String): Response<Unit>

    @PATCH("v1/tasks/{taskId}/completion")
    suspend fun toggleCompletion(
        @Path("taskId") taskId: String,
        @Body body: CompletionRequest,
    ): TaskDto

    // ── Projects ───────────────────────────────────────────────────

    @GET("v1/projects")
    suspend fun listProjects(): ItemsResponse<ProjectDto>

    @POST("v1/projects")
    suspend fun createProject(@Body body: ProjectCreateRequest): ProjectDto

    @PATCH("v1/projects/{id}")
    suspend fun updateProject(@Path("id") id: String, @Body body: ProjectCreateRequest): ProjectDto

    @DELETE("v1/projects/{id}")
    suspend fun deleteProject(@Path("id") id: String): Response<Unit>

    // ── Tags ───────────────────────────────────────────────────────

    @GET("v1/tags")
    suspend fun listTags(): ItemsResponse<TagDto>

    @POST("v1/tags")
    suspend fun createTag(@Body body: TagCreateRequest): TagDto

    @PATCH("v1/tags/{id}")
    suspend fun updateTag(@Path("id") id: String, @Body body: TagCreateRequest): TagDto

    @DELETE("v1/tags/{id}")
    suspend fun deleteTag(@Path("id") id: String): Response<Unit>
}
